README
########################## PNGwriter #########################################
#
#       Website: Main:             http://pngwriter.sourceforge.net/
#                Sourceforge.net:  http://sourceforge.net/projects/pngwriter/
#                Freshmeat.net:    http://freshmeat.net/projects/pngwriter/
#
#       Author:                    Paul Blackburn
#                                  Axel Huebl https://github.com/ax3l
#
#       Email:                     individual61@users.sourceforge.net
#
#       Version:                   0.5.5 (May 2014)
#
#       Description:               Library that allows plotting a 48 bit
#                                  PNG image pixel by pixel, which can
#                                  then be opened with a graphics program.
#
#       License:                   GNU General Public License
#                                  (C) 2002, 2003, 2004, 2005, 2006,
#                                  2007, 2008, 2009 Paul Blackburn
#                                  (C) 2013, 2014 Axel Huebl
##########################################################################


PNGwriter
Copyright 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009 Paul Blackburn
Copyright 2013, 2014 Axel Huebl



#######################################################################
                     IMPORTANT - IMPORTANTE
The  Spanish README and documentation can be found in doc/espaniol
El archivo LEAME y la documentacion en castellano esta en doc/espaniol
#######################################################################


#######################################################################
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#######################################################################




CONTACTS
Author's Email:                      individual61@users.sourceforge.net
Home page:                           http://pngwriter.sourceforge.net/
Sourceforge.net project page:        http://sourceforge.net/projects/pngwriter/
Freshmeat.net project page:          http://freshmeat.net/projects/pngwriter/
GitHub.com project page:             https://github.com/ax3l/pngwriter




INTRODUCTION
Hi. Thanks for downloading PNGwriter. I hope you find it
useful. See doc/english/CHANGES to see what's new.




DESCRIPTION
PNGwriter is a very easy to use open source graphics library that uses PNG 
as its output format. The interface has been designed to be as simple and 
intuitive as possible. It supports plotting and reading in the RGB 
(red, green, blue), HSV (hue, saturation, value/brightness) and CMYK 
(cyan, magenta, yellow, black) colour spaces, basic shapes, scaling, bilinear 
interpolation, full TrueType antialiased and rotated text support, bezier curves, 
opening existing PNG images and more. Documentation in English and Spanish. 
Runs under Linux, Unix, Mac OS X and Windows. Requires libpng and optionally 
FreeType2 for the text support.




REQUIREMENTS
LibPNG (http://www.libpng.org/)
PNGwriter requires a properly installed libpng in order to work.

Freetype 2 (http://www.freetype.org)
PNGwriter uses FreeType 2 to render antialiased text with kerning, but
it can be compiled without support for this. See below.

Fonts
If you are going to use PNGwriter's plot_text() function to plot text,
then you'll need a TrueType font file. PNGwriter installs a few from
the Free UCS Outline Fonts Project 
(http://savannah.nongnu.org/projects/freefont) in
/usr/local/share/pngwriter/fonts to get you started, or under whatever 
directory you chose with PREFIX at install time. See below. 




INSTALLATION
Before you do anything, check the 'make.include' file and make sure it is a
symbolic link to either one of 'make.include.linux' or
'make.include.osx'. 

The current setting to 'make.include.linux' should cover most systems,
including OS X. See below.

If you have an OS X system with libpng and FreeType2 installed via
Fink in /sw, or via MacPorts in /opt, edit and then use
'make.include.osx'. It is generally not necessary to use
make.include.osx on an OS X system if you are compiling with FreeType2
support. This is because the FreeType2 helper program freetype-config,
called at compile time, will actually provide the correct library and
header paths for your Fink or MacPorts distribution. In short, if
you're on OS X, just give make.include.linux a try, and if it doesn't
work, then tinker with make.include.osx


If you are using an old compiler, you may want to use
'make.include.linux.oldcpp', or alter your own makefile by adding
-DOLD_CPP to the compiler flags. 

If you wish to compile PNGwriter without FreeType2 support, then open
make.include with your text editor and uncomment the line that says

    # P_FREETYPE = 1

or, when compiling, add P_FREETYPE=1 to make's arguments.

Note that every time you compile a project that uses PNGwriter, and
PNGwriter has been compiled without FreeType2 support, you will have to
add -DNO_FREETYPE to your compilation flags.

To compile PNGwriter, just type

    make
   
and then, as Root, 

    make install
    
If you are not root, or do not want to install in the default
location (/usr/local/, set in make.include), then specify the location
with PREFIX when calling make, like this:

     make PREFIX=$HOME (for example)

or

     make PREFIX=$HOME/programming (for example)

The default installation locations are:

 -  libpngwriter.a in /usr/local/lib
 -  pngwriter.h in /usr/local/include
 -  The documentation in /usr/local/share/doc/pngwriter/doc/
 -  The examples in /usr/local/share/doc/pngwriter/examples/
 -  A few fonts in /usr/local/share/pngwriter/fonts

After installation, the following lines will list the installed
location of the software (determined from whatever PREFIX was at
install time):

 -  libpngwriter.a: /usr/local/lib/ 
 -  pngwriter.h: /usr/local/include/ 
 -  documentation: /usr/local/share/doc/pngwriter/ 
 -  examples: /usr/local/share/doc/pngwriter/ 
 -  fonts: /usr/local/share/pngwriter/fonts/

Look in examples/ for two examples of PNGwriter's use, and in
doc/english/EXAMPLES for an explanation.




SUPPORT
For examples, a FAQ, etc, take a look at the PNGwriter Home page:

http://pngwriter.sourceforge.net/

If you have a problem or a suggestion, you can use the support forum at

http://sourceforge.net/projects/pngwriter/

You can also email me at the address shown in the header. I would really like to hear
from you and what you are using PNGwriter for.




LEGAL
This library and its code are distributed under the GNU General
Public License. The complete text of the license is included in this
distribution. 




UPDATES
New versions will appear from time to time so check the web page.

http://pngwriter.sourceforge.net/
and
http://sourceforge.net/projects/pngwriter/

You can also stay updated by checking PNGwriter's freshmeat.net page:

http://freshmeat.net/projects/pngwriter/




WHAT'S IN THIS RELEASE? 

pngwriter-x.y.z
|-- Makefile
|-- README
|-- configure
|-- doc
|   |-- english
|   |   |-- CHANGES
|   |   |-- EXAMPLES
|   |   |-- LICENSE
|   |   |-- PNGwriterQuickReference_EN.pdf
|   |   `-- README
|   `-- espaniol
|       |-- CAMBIOS
|       |-- EJEMPLOS
|       |-- LEAME
|       |-- LICENCIA
|       `-- PNGwriterQuickReference_ES.pdf
|-- examples
|   |-- Makefile
|   |-- burro.png
|   |-- lyapunov.cc
|   |-- lyapunov.espaniol.cc
|   |-- pngtest.cc
|   `-- pngtest.espaniol.cc
|-- fonts
|   |-- FreeMonoBold.ttf
|   `-- FreeSansBold.ttf
|-- make.include -> make.include.linux
|-- make.include.linux
|-- make.include.linux.oldcpp
|-- make.include.osx
`-- src
    |-- Makefile
    |-- pngwriter.cc
    `-- pngwriter.h
    
    
    
    
THANKS

In no particular order, thanks to

> Xavier Andrade, Debian Package Maintainer for version 0.3.6
> Andres Kievsky
> Jorgen Pehrson, for the settext() function.
> Nadav Rotem, for suggesting the bezier() function.
> Jeramy Webb (jeramyw@gmail.com), for his generosity.
> Mike Heller (mkheller@gmail.com), for his generosity.
> Benjamin Raskob, for bringing to my attention the need for a resize() function.
> Dr John Charlery, for his continued support with the Windows-specific documentation.
> Tobias Kretz, for pointing out a bug in readfromfile(), and
> Greg Roelofs and John Bowler on png-implement for their help fixing the bug.
> Gurkan Sengun (gurkan@linuks.mine.nu, http://www.linuks.mine.nu/)
  for the filledtriangle() code.
> Miguel Gea (debian@miguelgea.com), Debian Package Maintainer for 
  versions > 0.5.1
> Carsten Klapp (carstenklapp@users.sourceforge.net), for his help
  with the Fink package.
> Sven, who posted anonymously on the PNGwriter forum
  concerning readfromfile()
> Mikkel (mikkel@biosource.dk) for expanding on Sven's code, greatly
  increasing the variety of PNGs that PNGwriter can open.

Have fun and tell me what you're up to!

Paul Blackburn
individual61@users.sourceforge.net


NEWS.md
# Latest News

## September 2017

Update on our last comment:
Since the policy change on SourceForge.net the adware issue on downloads for
unmaintained software was changed (about a year ago by now).

https://www.reddit.com/r/sysadmin/comments/4n3e1s/the_state_of_sourceforge_since_its_acquisition_in/

Great to hear!
(Still, we continue to stay on GitHub for now.)

## New version and a new team member — August 2015 Update

A long time has passed since PNGwriter's last release. Thanks to the skilled
and generous help of Axel Huebl ([@ax3l](http://github.com/ax3l)) many bugs
have been fixed, a new compilation procedure is in place, and PNGwriter is now
on [GitHub](https://github.com/pngwriter/pngwriter). Axel is now an active and
invaluable member of the PNGwriter Team, joining Paul Blackburn
([@individual61](https://github.com/individual61)), who created PNGwriter in
2002. Welcome aboard, Axel!

This nudge to action could not have come at a better time, since the
once-respectable host to the PNGwriter project, SourceForge.net, is no longer
what we consider a positive contribution to the Open Source community.
The PNGwriter website and files currently hosted on SourceForge.net will remain
on SourceForge.net, but the current and all future releases will be via
PNGwriter's GitHub repository.
 **Don't forget**!

Thank you, PNGwriter users, for your patience, bug fixes, and suggestions. I
hope you are as excited as we are to see where the next thirteen years take
PNGwriter!

Paul Blackburn  
San Jose, California  
August, 2015


tests/diamond.cc
/** Copyright 2011      Kevin Ryde
  * Copyright 2013-2014 Axel Huebl
  *
  * Example and bug report from Kevin Ryde
  *
  * License: GPLv2+
  *
  * http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=633405
  */

#include <pngwriter.h>
#include <cstdlib>

int main( )
{
  pngwriter img( 40,11, 0, "/tmp/foo.png" );

  img.diamond      ( 10,6, 8,8,  1.0,1.0,1.0 );
  img.filleddiamond( 30,6, 8,8,  1.0,1.0,1.0 );
  img.write_png();

  system( "convert /tmp/foo.png /tmp/foo.xpm; cat /tmp/foo.xpm" );
  return 0;
}


tests/readwrite.cc
/** Copyright 2016 Maximilian Knespel
  *
  * License: GPLv2+
  *
  */

#include <iostream>
#include <iomanip>
#include <cstdlib>      // srand, rand RAND_MAX
#include <cassert>
#include <cstring>      // memcpy
#include <pngwriter.h>

unsigned int const nMaxPixelsPerDimension = 2048;
unsigned int const nImageDimsToTest = 20;
char const * const tmpFile = "tmp-jhuio87er9r.png";

void writeArray
(
    float * const rData,
    unsigned int const rWidth,
    unsigned int const rHeight
)
{
    pngwriter png( rWidth,rHeight, 0, tmpFile );

    for ( unsigned int ix = 0; ix < rWidth ; ++ix )
    for ( unsigned int iy = 0; iy < rHeight; ++iy )
    {
        float value = rData[ iy*rWidth + ix ];
        png.plot( 1+ix,1+iy, value, value, value );
    }

    png.close();
}


void readArray
(
    float * const rData,
    unsigned int const rWidth,
    unsigned int const rHeight
)
{
    pngwriter png( 1,1, 0 /* background color */, tmpFile );
    png.readfromfile( tmpFile );
    assert( png.getwidth () == (int) rWidth  );
    assert( png.getheight() == (int) rHeight );

    for ( unsigned int ix = 0; ix < rWidth ; ++ix )
    for ( unsigned int iy = 0; iy < rHeight; ++iy )
    {
        double r = png.dread( 1+ix,1+iy, 1 );
        double g = png.dread( 1+ix,1+iy, 1 );
        double b = png.dread( 1+ix,1+iy, 1 );
        assert( r == g and g == b );

        rData[ iy*rWidth + ix ] = r;
    }

    png.close();
}


int readWriteImage
(
    unsigned int const rWidth,
    unsigned int const rHeight
)
{
    /* initialize test data we want to write out and then read in again */
    float * pLastValues  = new float[ rWidth * rHeight ];
    float * pValues      = new float[ rWidth * rHeight ];
    for ( unsigned int i = 0; i < rWidth * rHeight; ++i )
        pLastValues[i] = (float) rand() / RAND_MAX;

    /* First write out value and read it in again and ignore tests, because
     * the first read/write cycle brings almost definitely different values.
     * But the next cycles shouldn't anymore or else there is no guarantee
     * the values being read and written will stop changing */
    writeArray( pLastValues, rWidth, rHeight );
    readArray ( pLastValues, rWidth, rHeight );

    writeArray( pLastValues, rWidth, rHeight );
    readArray ( pValues    , rWidth, rHeight );

    for ( unsigned int ix = 0; ix < rWidth ; ++ix )
    for ( unsigned int iy = 0; iy < rHeight; ++iy )
    {
        unsigned int const i = iy*rWidth + ix;
        assert( pLastValues[i] == pValues[i] );
    }

    delete[] pLastValues;
    delete[] pValues;

    return 0;
}


int testReadWrite( void )
{
    for ( unsigned int i = 0; i < nImageDimsToTest; ++i )
    {
        unsigned int width  = rand() % nMaxPixelsPerDimension;
        unsigned int height = rand() % nMaxPixelsPerDimension;
        if ( width  == 0 ) width  = 1;
        if ( height == 0 ) height = 1;
        std::cout << "Doing read write test with image "
                  << width << " x " << height << " pixels large\n";
        int error = readWriteImage( width, height );
        if ( error != 0 )
            return error;
    }
    return 0;
}


int testWithOnePixel( void )
{
    unsigned int const nValuesToTest = 1000;
    unsigned int iValuesFlawed = 0;
    srand(16546485);
    for ( unsigned int i = 0; i < nValuesToTest; ++i )
    {
        float firstValue = (float) rand() / RAND_MAX;
        writeArray( &firstValue, 1,1 );
        readArray ( &firstValue, 1,1 );

        float lastValue = firstValue;
        int iRepetition = 0;
        for ( iRepetition = 0; iRepetition < 100000; ++iRepetition )
        {
            float value;
            writeArray( &lastValue, 1,1 );
            readArray ( &value    , 1,1 );

            if ( lastValue == value )
                break;
            lastValue = value;
        }
        if ( iRepetition > 0 )
        {
            std::cout << "Input Value "   << std::setw(9) << firstValue
                      << " converged to " << std::setw(9) << lastValue
                      << " in "           << std::setw(5) << iRepetition
                      << " iterations" << std::endl;
            ++iValuesFlawed;
        }
    }
    std::cout << "Found " << iValuesFlawed << " problematic values out of "
              << nValuesToTest << "\n";

    return iValuesFlawed == 0;
}


int main( void )
{
    srand(16546485);
    int error = 0;
    error &= testWithOnePixel();
    error &= testReadWrite();
    return error;
}


tests/blackwhite.cc
/** Copyright 2014-2016 Felix Schmitt, Axel Huebl
  *
  * Example and bug report from Felix Schmitt
  *
  * License: GPLv2+
  *
  * https://github.com/pngwriter/pngwriter/issues/13
  */

#include <pngwriter.h>
#include <iostream>

int main(int argc, char **argv)
{
    if (argc < 2)
        return 1;

    const std::string filename = argv[1];

    const size_t size_x = 256, size_y = 256;

    pngwriter image(size_x, size_y, 0, (filename + std::string(".tmp")).c_str());
    image.readfromfile(filename.c_str());

    size_t width = image.getwidth();
    size_t height = image.getheight();
    size_t bit_depth = image.getbitdepth();

    bool black = false, white = false;
    bool black_int = false, white_int = false;

    std::cout << "bit depth = " << bit_depth << std::endl;

    for (size_t x = 0; x < width; ++x)
        for (size_t y = 0; y < height; ++y)
        {
            /* pngwriter coordinates start at (1,1) and the y direction is inverted */
            int pos_x = 1 + x;
            int pos_y = height - y - 1;

            double color = image.dread(pos_x, pos_y,3);
            std::cout << "dread val = " << color << std::endl;
            if (color == 0.0)
                black = true;

            if (color == 1.0)
                white = true;

            int col = image.read(pos_x, pos_y);//,3);
            std::cout << "read val = " << col << std::endl;
            if (col == 0)
                black_int = true;

            // bit_depth 8 and 16 get all read as 16bit
            if (col == 65535)
                white_int = true;
        }

    image.close();

    if (!(black && white))
    {
        std::cout << "failed double" << std::endl;
        return 1;
    }

    if (!(black_int && white_int))
    {
        std::cout << "failed int" << std::endl;
        return 1;
    }

    return 0;
}


examples/blackwhite.cc
/** Copyright 2014 Felix Schmitt, Axel Huebl
  *
  * Example and bug report from Felix Schmitt
  *
  * License: GPLv2+
  *
  * https://github.com/ax3l/pngwriter/issues/13
  */

#include <pngwriter.h>
#include <iostream>

int main(int argc, char **argv)
{
    if (argc < 2)
        return 1;

    const std::string filename = argv[1];

    const size_t size_x = 256, size_y = 256;

    pngwriter image(size_x, size_y, 0, (filename + std::string(".tmp")).c_str());
    image.readfromfile(filename.c_str());

    size_t width = image.getwidth();
    size_t height = image.getheight();
    size_t bit_depth = image.getbitdepth();

    bool black = false, white = false;
    bool black_int = false, white_int = false;

    std::cout << "bit depth = " << bit_depth << std::endl;

    for (size_t x = 0; x < width; ++x)
        for (size_t y = 0; y < height; ++y)
        {
            /* pngwriter coordinates start at (1,1) and the y direction is inverted */
            int pos_x = 1 + x;
            int pos_y = height - y - 1;

            double color = image.dread(pos_x, pos_y,3);
            //std::cout << "dread val = " << color << std::endl;
            if (color == 0.0)
                black = true;

            if (color == 1.0)
                white = true;

            int col = image.read(pos_x, pos_y,3);
            //std::cout << "read val = " << col << std::endl;
            if (col == 0)
                black_int = true;

            if (col == 65535 && bit_depth == 16)
                white_int = true;
            if (col == 255 && bit_depth == 8)
                white_int = true;
        }

    image.close();

    if (!(black && white))
        std::cout << "failed double" << std::endl;

    if (!(black_int && white_int))
        std::cout << "failed int" << std::endl;

    return 0;
}


examples/pngtest.cc
/********************************* PNGwriter **********************************
*
*   Website: Main:             http://pngwriter.sourceforge.net/
*            GitHub.com:       https://github.com/pngwriter/pngwriter
*            Sourceforge.net:  http://sourceforge.net/projects/pngwriter/
*
*
*    Author:                    Paul Blackburn https://github.com/individual61
*                               Axel Huebl https://github.com/ax3l
*
*    Email:                     individual61@users.sourceforge.net
*
*    Version:                   0.6.0 (March 2017)
*
*    Description:               Library that allows plotting a 48 bit
*                               PNG image pixel by pixel, which can
*                               then be opened with a graphics program.
*
*    License:                   GNU General Public License
*                               (C) 2002-2017 Paul Blackburn
*                               (C) 2013-2017 Axel Huebl
*
******************************************************************************/

/*
 * ######################################################################
 *  This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be
 * useful, but WITHOUT ANY WARRANTY; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 * ######################################################################
 * */

#include <pngwriter.h>
#include <cmath>
#include <iostream>
#include <string>
using namespace std;
#include <ctime>
#include <cstdlib>
#include <cstdio>

bool file_exists( const char * const name )
{
   FILE *file = fopen( name, "r" );
   if ( file )
   {
      fclose(file);
      return true;
   }
   else
   {
      return false;
   }
}

int main()
{
   /* Introduction
    * This program is meant to serve as an introduction to using PNGwriter.
    * It is not the simplest possible use of it, and does not cover all
    * of its features. See the website for more examples.
    * This program will generate 4 PNG images. The first two will be
    * examples of using PNGwriter's figures.
    * The third will be an example of how to read an already existing PNG image.
    * The fourth will be an example of how to use plotHSV to generate a rainbow.
    * */

   //////////////////////////////////////////////////////////////////////

   /* one.png
    * This will be a 300x300 image with a black background, called one.png
    * */
   pngwriter one(300,300,0,"one.png");

   std::cout << "Generating one.png...";

   /* Purple rectangle (filled)
    * Make a purple filled retangle. Notice we are using
    * int colour levels, e.g. 65535 is the max level of green.
    * */
   one.filledsquare(30,5,45,295,65535,0,65535);

   /* Blue rectangle
    * Make a blue rectangle over the green one.
    * Notice that we are using colour coefficcients of type double.
    * */
   one.square(20,10,40,290,0.0,0.0,1.0);

   /* Yellow circle (filled) and Grey Circle
    * Using colour coefficcients of type double.
    * */
   one.filledcircle(250,250,40,1.0,1.0,0.0);
   one.circle(250,70,30,20000,20000,50000);

   /* Lines
    * Draw 16 lines to test all general directions of line().
    * All start from the centre and radiate outwards.
    * The colour is varied to be able to distinguish each line.
    * */
   //Top row
   one.line(150,150,0,300,0,65535,4000);
   one.line(150,150,75,300,0,65535,8000);
   one.line(150,150,150,300,0,65535,12000);
   one.line(150,150,225,300,0,65535,16000);
   one.line(150,150,300,300,0,65535,20000);
   //Right side
   one.line(150,150,300,225,0,65535,24000);
   one.line(150,150,300,150,0,65535,28000);
   one.line(150,150,300,75,0,65535,32000);
   one.line(150,150,300,0,0,65535,36000);
   //Bottom row
   one.line(150,150,225,0,0,65535,40000);
   one.line(150,150,150,0,0,65535,44000);
   one.line(150,150,75,0,0,65535,48000);
   one.line(150,150,0,0,0,65535,52000);
   //Left side
   one.line(150,150,0,75,0,65535,56000);
   one.line(150,150,0,150,0,65535,60000);
   one.line(150,150,0,225,0,65535,64000);

   /*Change the text information in the PNG header
    * */
   one.settext("one.png", "John Cleese", "Test PNG", "pngtest");

   std::cout << " done. Writing to disk...";
   one.close();
   std::cout << " done." << std::endl;


  ////////////////////////////////////////////////////////////////////////////////////
   /*  two.png
    *  This will be a 300x300 image with a black background, it will be called two.png.
    * Note that we are using 0.0 as the background colour and that we are using a string
    * type object as the filename, which can convert itself into a const char * with
    * filename.c_str().
    * */
   std::cout << "Generating two.png...";
   string filename = "two.png";
   pngwriter two(300,300,0.0,filename.c_str());

   /* Gradients
    * Here we will draw a few gradients, which are nothing more than drawing a filledsquare
    * or line or filledcircle repeatedly, changing the colour as we go.
    * */

   /* Draw the gradient in the lower left corner.
    * */
   for(int iter = 1;iter< 300;iter++)
     {
	two.line(1,300,iter,1,65535-int(65535*((double)iter)/300.0), 0,65535);
     }

   /* Draw the green circles.
    * */
   for(int iter2 = 0; iter2 < 8; iter2++)
     {
	two.filledcircle(280-25*iter2,250,10,0,65535 - int(65535*((double)iter2/7.0)),0);
     }

   /* Draw the other central gradient.
    * Notice that it is drawing a filled square, and moving the upper right corner down
    * and left, while changind the colour.
    * */
   for(int iter3 = 120; iter3>0;iter3--)
     {
	two.filledsquare(70,70,70+iter3,70+iter3,0,0,65535 -int(65535*((double)iter3/120.0 )))	;
     }

   /* Draw the orange circle gradient.
    * */
   for(int iter4 = 0; iter4<31; iter4++)
     {
	two.filledcircle(250,200,30-iter4,1.0,1.0 - double(iter4)/30.0,0.0);
     }


   /* Now, just as an example, we will use a lower compression on this image.
    * The default is 6 (from 0 to 9) and we will set it to 3. The lower the compression used
    * the faster the image will be close()d. Complex images will take longer to
    * close() than simple ones.
    * */
   two.setcompressionlevel(3);
   std::cout << " done. Writing to disk...";
   two.close();
   std::cout << " done." << std::endl;


   ///////////////////////////////////////////////////////////////////

   /* copiaburro.png
    * This section opens burro.png and places it in the pngwriter instance called burro.
    * The PNGwriter instance burro is 1x1 pixels in size because we will use it
    * as a container for another PNG image that we will
    * soon read in. No point in grabbing memory now, so we'll set the width and height to 1.
    * Notice that the original is not altered in any way.
    * */
   pngwriter burro(1,1,0,"copiaburro.png");

   /* readfromfile()
    * Now we specify the path to the file we wish to place into
    * the PNGwriter instance "burro":
    * */
   std::string burro_path = std::string( "burro.png" );
   if ( not file_exists( burro_path.c_str() ) )
      burro_path = std::string( "../examples/burro.png" );
   if ( not file_exists( burro_path.c_str() ) )
      burro_path = std::string( "./examples/burro.png" );
   std::cout << "Opening '" << burro_path << "'...";
   burro.readfromfile( burro_path.c_str() ); //It really is that easy.
   std::cout << " done." << std::endl;

   /* getwidth() and getheight()
    * We may not know the size of the image we just read in.
    * We find out like this:
    * */
   int burrowidth = burro.getwidth();
   int burroheight = burro.getheight();


   std::cout << "The image that has just been read from disk (burro.png) is " << burro.getheight();
   std::cout << " pixels high and " << burro.getwidth()<< " pixels wide." << std::endl;
   std::cout << "Bit depth is " << burro.getbitdepth() << std::endl;
   std::cout << "Image gamma is: " << burro.getgamma() << std::endl;

   /* Cover the Donkey
    * We now draw a red circle over the donkey, just to prove that the image was read in correctly,
    * and that it was manipulated.
    * */
   burro.filledcircle((int)(burrowidth/3.0),(int)(2.0*burroheight/3.0),(int)(burrowidth/5),65535,0,0);

   /* close()
    * Must close this instance to have it written to disk.
    * */
   std::cout << "Writing to disk...";
   burro.close();
   std::cout << " done." << std::endl;

   /////////////////////////////////////////////////////////////////////

   /* Rainbow
    * To demonstrate the use and usefulness of plotHSV(), we will
    * generate a rainbow.
    * We will use the function plotHSV that
    * plots a pixel at x, y with the colour given in the HSV colourspace
    * (hue, saturation, value), instead of the traditional RGB.
    * */

   std::cout << "\nCreating arcoiris.png...";
   pngwriter arcoiris(300,300,0,"arcoiris.png");
   std::cout << " done." <<endl;
   std::cout << "Plotting...";
   for(int a= 1;a<301;a++)
      for(int b= 1;b<301;b++)
       arcoiris.plotHSV(a,b,double(a)/300.0,double(b)/300.0,1.0);

   std::cout << " done." << endl;

   arcoiris.setgamma(0.5);
   std::cout << "Writing to disk...";
   arcoiris.close();
   std::cout << " done." << std::endl;
   /////////////////////////////////////////////////////////////////

   /* Random filled blended triangles.*/

   srand ( time(NULL) );

   std::cout << "\nCreating triangles.png...";
   pngwriter triangles(400,400,0,"triangles.png");
   std::cout << " done.";
   double opacityrange = 0.3;

   std::cout << "\nPlotting triangles...";
   for(int jj = 0; jj < 10;  jj++)
     {

	//This creates the blue triangle outlines.
	triangles.triangle(
				       ((int) rand()%500) - 50,
				       ((int) rand()%500) - 50,
				       ((int) rand()%500) - 50,
				       ((int) rand()%500) - 50,
				       ((int) rand()%500) - 50,
				       ((int) rand()%500) - 50,
				       0.0, 0.0, 1.0);
     }

   for(int ii = 0; ii < 20;  ii++)
     {
	//This creates the red blended triangles
	triangles.filledtriangle_blend(
				       ((int) rand()%500) - 50,
				       ((int) rand()%500) - 50,
				       ((int) rand()%500) - 50,
				       ((int) rand()%500) - 50,
				       ((int) rand()%500) - 50,
				       ((int) rand()%500) - 50,
				       0.4 + ((double)(rand()/RAND_MAX))*opacityrange,
				       1.0, 0.0, 0.0);
     }
   std::cout << " done.";
   std::cout << "\nWriting to disk...";
   triangles.close();
   std::cout << " done.";
   std::cout << "\n\npngtest has finished. Take a look at the PNG images that have been created!\n";
   return 0;
}
